packages/framework/undo-redo/src/sequenceHandler.ts

```diff
@@ -85,10 +89,22 @@ export class SharedSegmentSequenceRevertible implements IRevertible {
                     };
                     this.tracking.push(current);
                 }
+
+                if (event.deltaOperation === MergeTreeDeltaType.REMOVE) {
+                    this.addRemovedRefs(current, range.segment);
+                }
             }
         }
     }
 
+    private addRemovedRefs(current: ITrackedSharedSegmentSequenceRevertible, segment: ISegment): void {
+        const segRefs = Array.from(segment.localRefs ?? [], (ref) => ({ ref, offset: ref.getOffset() }));
+        if (segRefs.length > 0) {
+            current.refsToRestore ??= new Map();
+            current.refsToRestore.set(segment, segRefs);
```

Comment:
i don't think this works, as the segment could still split after being removed, as it is still above the min seq, that could then mess with the mapping.

Reply:
ah, yeah--that makes sense. It should be possible to do something similar but using tracking groups, right? (haven't looked too much into the details, but I see that they split how one would want on segment split)

Then--is that logic something worth exporting from MergeTree? At first glance it looks like it should be possible to write in a tree-shakeable way, so hopefully no impact for non-users.

Reply:
alternatively, seems like `mergeTreeMaintenanceCallback` has enough information to remedy this. It's a lot of burden to put on the implementer though.

Reply:
I pushed a change with a test for this kind of thing (and haven't tweaked implementation, so it fails). Going to table the undo/redo part of this PR for now and split out the interval collection changes into a separate thing.